### PR TITLE
Adding diagram display utilities to `chalk`

### LINKED ISSUE #44: Add diagram display utilities, with or without saving image files on disk
- [x] Add a new `chalk.utils` module. This module will offer various diagram display utilities as follows.

  - [x] `utils.show(filepath: str)`: 
    > Use this function for displaying saved images from disk.
  
  - [x] `utils.imgen(d: Diagram, **kwargs)`: 
    > Use this function to display a diagram, without creating any permanent file on disk. 
    > :point_right: This function does the following:
    > 1. Creates a temporary file.
    > 2. If necessary, it also creates a temporary folder for storing the temporary file.
    > 3. Displays the temporary-image-file for a user-specified (optional) `5` seconds.
    > 4. Clears up both the temporary file and folder after the 5 seconds have elapsed.

  - [x] `utils.quick_probe(**kwargs)`:
    > Use this function to **quickly generate a default sample Diagram** (or a user-provided Diagram) and **display it _without saving it_**.

I believe, with these utilities it will be quite easy to visualize diagrams without having to save them on the disk.

### LINKED ISSUE #46: Add docstrings to `chalk.core.Diagram`
Add docstring to all the methods under `chalk.core.Diagram`.

```python
import chalk
print('\n'.join([f"- [x] `{e}()`" for e in dir(chalk.core.Diagram) if not e.startswith("_")]))
```

The methods of **chalk.core.Diagram**:

Legend: :orange_circle:  Not Applicable 

- [x] `above()`
- [x] `above2()`
- [x] `align_b()`
- [x] `align_bl()`
- [x] `align_br()`
- [x] `align_l()`
- [x] `align_r()`
- [x] `align_t()`
- [x] `align_tl()`
- [x] `align_tr()`
- [x] `apply_transform()`
- [x] `at_center()`
- [x] `atop()`
- [x] `beside()`
- [x] `center_xy()`
- [x] `dashing()`
- [x] `display()`
- [x] `fill_color()`
- [x] `fill_opacity()`
- [x] `get_bounding_box()`
- [x] `get_subdiagram_bounding_box()` :orange_circle: 
- [x] `line_color()`
- [x] `line_width()`
- [x] `named()`
- [x] `pad()`
- [x] `pad_b()`
- [x] `pad_l()`
- [x] `pad_r()`
- [x] `pad_t()`
- [x] `reflect_x()`
- [x] `reflect_y()`
- [x] `render()`
- [x] `render_svg()`
- [x] `rotate()`
- [x] `rotate_by()`
- [x] `scale()`
- [x] `scale_uniform_to_x()`
- [x] `scale_uniform_to_y()`
- [x] `scale_x()`
- [x] `scale_y()`
- [x] `shear_x()`
- [x] `shear_y()`
- [x] `show_bounding_box()`
- [x] `show_origin()`
- [x] `to_list()`
- [x] `to_svg()` :orange_circle: 
- [x] `translate()`
- [x] `translate_by()`

### LINKED ISSUE #47: Add a display method to `chalk.core.Diagram`
Adding a display method to `chalk.core.Diagram` object could be very helpful in visualizing the diagram, without having to save it on disk.

The `chalk.utils.imgen()` function will be used here.

### LINKED ISSUE #48: Add docstrings to all methods under `chalk.bounding_box.BoundingBox`
Add docstring to all the methods under `chalk.bounding_box.BoundingBox`.

```python
import chalk
print('\n'.join([f"- [x] `{e}()`" for e in dir(chalk.bounding_box.BoundingBox) if not e.startswith("_")]))
```

The methods of **chalk.core.Diagram**:

- [x] `apply_transform()`
- [x] `bl()`
- [x] `bottom()`
- [x] `cardinal()`
- [x] `center()`
- [x] `empty()`
- [x] `enclose()`
- [x] `from_limits()`
- [x] `height()`
- [x] `left()`
- [x] `reflect_x()`
- [x] `reflect_y()`
- [x] `right()`
- [x] `rotate()`
- [x] `rotate_by()`
- [x] `scale()`
- [x] `scale_x()`
- [x] `scale_y()`
- [x] `shear_x()`
- [x] `shear_y()`
- [x] `top()`
- [x] `tr()`
- [x] `translate()`
- [x] `translate_by()`
- [x] `union()`
- [x] `width()`

### LINKED ISSUE #49: Add docstrings to `chalk.point.Vector`
Add docstring to all the methods under `chalk.point.Vector`.

```python
import chalk
print('\n'.join([f"- [x] `{e}()`" for e in dir(chalk.point.Vector) if not e.startswith("_")]))
```

The methods of **chalk.point.Vector**:

- [x] `angle()`
- [x] `apply_transform()`
- [x] `from_polar()`
- [x] `length()`
- [x] `reflect_x()`
- [x] `reflect_y()`
- [x] `rotate()`
- [x] `rotate_by()`
- [x] `scale()`
- [x] `scale_x()`
- [x] `scale_y()`
- [x] `shear_x()`
- [x] `shear_y()`
- [x] `translate()`
- [x] `translate_by()`

### LINKED ISSUE #50: Add docstrings to `chalk.point.Point`
Add docstring to all the methods under `chalk.point.Point`.

```python
import chalk
print('\n'.join([f"- [x] `{e}()`" for e in dir(chalk.point.Point) if not e.startswith("_")]))
```

The methods of **chalk.point.Point**:

- [x] `apply_transform()`
- [x] `reflect_x()`
- [x] `reflect_y()`
- [x] `rotate()`
- [x] `rotate_by()`
- [x] `scale()`
- [x] `scale_x()`
- [x] `scale_y()`
- [x] `shear_x()`
- [x] `shear_y()`
- [x] `translate()`
- [x] `translate_by()`

### LINKED ISSUE #51: Add docstrings to `chalk.trail.Trail`
Add docstring to all the methods under `chalk.trail.Trail`.

```python
import chalk
print('\n'.join([f"- [x] `{e}()`" for e in dir(chalk.trail.Trail) if not e.startswith("_")]))
```

The methods of **chalk.trail.Trail**:

- [x] `apply_transform()`
- [x] `from_path()`
- [x] `reflect_x()`
- [x] `reflect_y()`
- [x] `rotate()`
- [x] `rotate_by()`
- [x] `scale()`
- [x] `scale_x()`
- [x] `scale_y()`
- [x] `shear_x()`
- [x] `shear_y()`
- [x] `stroke()`
- [x] `to_path()`
- [x] `transform()`
- [x] `translate()`
- [x] `translate_by()`

### LINKED ISSUE #52: Add docstrings to all classes in `chalk/core.py`
Add docstring to all the classes other than `Diagram` in `chalk/core.py`.

- [x] `Primitive`
- [x] `Empty`
- [x] `Compose`
- [x] `ApplyTransform`
- [x] `ApplyStyle`
- [x] `ApplyName`

Note: `chalk.core.Diagram` is already covered by issue #47

### PR DESCRIPTION
**Pull request by sugatoray**
This PR adds the functionality of displaying a `chalk.Diagram` with a few convenience functions.

- [x] Adds the new `chalk.utils` module: Closes #44 

  - [x] `utils.show(filepath: str)`: 
    > Use this function for displaying saved images from disk.
  
  - [x] `utils.imgen(d: Diagram, **kwargs)`: 
    > Use this function to display a diagram, without creating any permanent file on disk. 
    > :point_right: This function does the following:
    > 1. Creates a temporary file.
    > 2. If necessary, it also creates a temporary folder for storing the temporary file.
    > 3. Displays the temporary-image-file for a user-specified (optional) `5` seconds.
    > 4. Clears up both the temporary file and folder after the 5 seconds have elapsed.

  - [x] `utils.quick_probe(**kwargs)`:
    > Use this function to **quickly generate a default sample Diagram** (or a user-provided Diagram) and **display it _without saving it_**.

  - [x] `create_sample_diagram()`:
    > Use this function to create a sample diagram, with multiple shapes, colors, etc. -- *with only a single function call*.

I believe, with these utilities it will be quite easy to visualize diagrams without having to save them on the disk.

- [x] Adds a `Diagram.display()` method: Closes #47
  > This facilitats convenient visualization of a diagram without saving any image file permanently on disk. 

- [x] Adds docstrings to methods of:
  - [x] `chalk.core.Diagram` class. Closes #46 
  - [x] `chalk.bounding_box.BoundingBox` class. Closes #48  
  - [x] `chalk.point.Point` class. Closes #50 
  - [x] `chalk.point.Vector` class. Closes #49 
  - [x] `chalk.trail.Trail` class. Closes #51 
  - [x] all the classes other than `Diagram` in `chalk/core.py`: Closes #52

### CODE REVIEW

**Comment by danoneata** [comment]
Thanks a lot for the contributions! I think that having a way of easily displaying diagrams can be very useful. But I'm not convinced of a couple of decisions:
- Having a timed display. I would find it more convenient to view the image until I decide to close it; I don't know if it's easy to clean the temporary file in this scenario.
- Displaying a default diagram (as done by `quick_probe`). I'm not really sure how useful is to display a default image, but maybe I'm missing some use case.

**Comment by danoneata** [comment]
Thanks a lot for the contributions! I think that having a way of easily displaying diagrams can be very useful. But I'm not convinced of a couple of decisions:
- Having a timed display. I would find it more convenient to view the image until I decide to close it; I don't know if it's easy to clean the temporary file in this scenario.
- Displaying a default diagram (as done by `quick_probe`). I'm not really sure how useful is to display a default image, but maybe I'm missing some use case.

**Comment by sugatoray** [comment]
> * Having a timed display. I would find it more convenient to view the image until I decide to close it; I don't know if it's easy to clean the temporary file in this scenario.

The image once displayed in a window after running `imgen()` will self-destruct in (say) 5 seconds. But the window will last until you close it. It's not what I had planned for, but I guess it's an unplanned side-effect, which for now, is useful and gives you exactly what you suggested. 

> * Displaying a default diagram (as done by `quick_probe`). I'm not really sure how useful is to display a default image, but maybe I'm missing some use case.

`quick_probe()` does two things:

- Offers to create a dummy `Diagram` and perform some actions if you would.
- Also, it offers a toned-down version of `imgen()` without all the bells and whistles `imgen()` comes with. You could also pass any `Diagram` to `quick_probe()` and see a temporary output.

There is a certain amount of functionality-overlap between `imgen()` and `quick_probe()`. But scenarios, when you are only interested in getting a dummy test diagram, after you have made some changes to the code base, this could be helpful.

@danoneata Please let me know your thoughts on this.

**Comment by sugatoray** [comment]
## Added convenient (fancy + colored) logging support with `loguru`

image

**Comment by sugatoray** [comment]
:fire: Displaying sample `Diagram`s conveniently with the new `.display()` method.

```python
from chalk.utils import create_sample_diagram
d = create_sample_diagram(option="a|b")
a, b = create_sample_diagram(option="a,b")

a.display(height=256)
b.display(height=256)
d.display(height=256)
```
| ![image-a][a] </br></br> <p> `a.display()` </p> | ![image-b][b] </br></br> <p> `b.display()` </p> | ![image-d][d] </br></br> <p> `d.display()` </p> |
|:---:|:---:|:---:|

[a]: https://user-images.githubusercontent.com/10201242/171998764-54e0b3b8-cbe3-4a53-82ab-d7439962ce95.png
[b]: https://user-images.githubusercontent.com/10201242/171998871-e94318e8-30e0-4818-b0dc-4c74fbcc05ad.png
[d]:https://user-images.githubusercontent.com/10201242/171998711-19e7c3ca-256c-4a20-b1f9-5310c881e97a.png

cc: @danoneata

**Comment by sugatoray** [comment]
## Added docstrings for all the methods of `chalk.core.Diagram`

The following is partial output of `make interrogate` command (only showing the `COVERED` functions/methods; omitted `MISSED` ones)

image

image


image

**Comment by danoneata** [comment]
Thanks a lot for taking a stab at the documentation! I will take a closer look next week. Just a quick question: do you use a specific format for the docstrings? For example, I see that you have used double backtikcs ` `` ` in the docstrings—I assume this is related to a monospace font?

> The image once displayed in a window after running `imgen()` will self-destruct in (say) 5 seconds. But the window will last until you close it. It's not what I had planned for, but I guess it's an unplanned side-effect, which for now, is useful and gives you exactly what you suggested.

I see, that sounds good! In this case, maybe the log should only output "Removed temporary image!" instead of "Closed and removed temporary image!"?

> But scenarios, when you are only interested in getting a dummy test diagram, after you have made some changes to the code base, this could be helpful.

Okay, I understand the motivation now. I guess I didn't feel the need of this behaviour since I was using the scripts in the `examples` folder as a way to test the code changes.

File: chalk/core.py [hunk 1]
```diff
@@ -111,6 +144,19 @@ def _repr_svg_(self) -> str:
         return svg
 
     def atop(self, other: "Diagram") -> "Diagram":
+        """Given to diagrams ``a`` and ``b``, ``a.atop(b)``
```
Comment (danoneata) [inline chalk/core.py:147]: "Given to" should be "Given t**w**o". This small typo also appears for other functions.

Comment (sugatoray) [reply chalk/core.py:147]: Thanks. I had written for one and copy pasted for the others. 🤦‍♂️🤦‍♂️

Also, on a separate note, please check this out. 

- https://github.blog/2022-04-22-5-simple-things-every-developer-can-do-to-ship-more-secure-code/

Comment (sugatoray) [reply chalk/core.py]: Made the corrections. Thank you for catching them.

image

**Comment by sugatoray** [comment]
> Thanks a lot for taking a stab at the documentation! I will take a closer look next week. Just a quick question: do you use a specific format for the docstrings? For example, I see that you have used double backtikcs `` `` `` in the docstrings—I assume this is related to a monospace font?

I have plans to add a documentation framework for the docs here. Typically, if you use ``something``, that would show up as expected (inline code) in the docs. The next PR I work on I will start adding the documentation framework.

---

> I see, that sounds good! In this case, maybe the log should only output "Removed temporary image!" instead of "Closed and removed temporary image!"?

Makes sense. But I would suggest specifying *"Removed temporary image **file**!"* (added _`file`_ to your suggestion).

---

> Okay, I understand the motivation now. I guess I didn't feel the need of this behaviour since I was using the scripts in the `examples` folder as a way to test the code changes.

:+1:

**Comment by sugatoray** [comment]
### Added docstring to `chalk.BoundingBox`

image

image

**Comment by sugatoray** [comment]
### Added docstring to `chalk.Point` and `chalk.Vector` classes

image

image

**Comment by sugatoray** [comment]
### Added docstring to `chalk.Trail`

image

image

**Comment by sugatoray** [comment]
### Added docstring to `chalk.core.X` where `X` is NOT `Diagram`

- #52

image

image

**Comment by sugatoray** [comment]
### Partially updated docstrings for `chalk.style.Style`

image

image

File: chalk/bounding_box.py [hunk 2]
```diff
@@ -6,30 +6,92 @@
...
+        """Returns the position of the bottom-left corner
+        of the bounding box.
+
+        Returns:
+            Point: A point object (``chalk.point.Point``).
+        """
         return Point(self.left, self.bottom)
 
     def cardinal(self, dir: str) -> Point:
+        """Returns the position of and edge or a corner of the bounding
```
Comment (danoneata) [inline chalk/bounding_box.py:70]: Small typo: "of an~d~ edge"

File: chalk/bounding_box.py [hunk 3]
```diff
@@ -44,35 +106,83 @@ def cardinal(self, dir: str) -> Point:
...
+        """
         return self.br.y
 
     @property
     def center(self) -> Point:
+        """Returns position of the center
+        of the bounding box.
+
+        Returns:
+            Point: A point object (``chalk.point.Point``).
```
[the remark this answers is not in the document]
Comment (sugatoray) [reply chalk/bounding_box.py:171]: At the time I was writing the docs, I kept it. But now after creating the docs website demo, I believe these could be dropped summarily.

[the remark this answers is not in the document]
Comment (sugatoray) [reply chalk/bounding_box.py]: Done. (removed those)

File: chalk/bounding_box.py (hunk 2 above)
Comment (sugatoray) [reply chalk/bounding_box.py]: Done.

**Comment by sugatoray** [comment]
All suggested changes have been made.

cc: @danoneata

**Comment by sugatoray** [comment]
Hi @danoneata, 

Please let me know if any other change you would propose. Otherwise, if it looks good, please merge it. I will then merge the master branch into the other PR of mine, which was branched from the PR branch of this PR -- and hence it seems a lot of changes have been made there. The other PR mostly only adds the docs infrastructure on top of this and adds some documentation to some parts of the code base.

The idea here is to get the docs site up and running so all contributors can add more meaningful/rich documentation as and when you please. 

Thank you.

**Comment by srush** [comment]
I'll merge this so that you are no longer blocked. Please do try to separate these out better in the future though. I am not sure what is different about this PR and the other one?

**Comment by danoneata** [comment]
Thanks a lot for helping with this PR @srush!

I also feel that there are too many things going on in this PR and I am somewhat overwhelmed by all the changes.

Strictly speaking regarding the display utilities from `utils.py`, I know that a lot of work has gone into this and I appreciate that, but I'm still having some mixed feelings:

- I'm still not sold on the utility of the `create_sample_diagram` or the `quick_probe` functions; if we want to produce some default diagrams for testing purposes aren't the ones in the `examples` folder enough?
- I'm not sure about the logging part in `imgen` (it's a bit too verbose for my taste) and it's unclear whether adding a new dependency (`loguru`) just for this is justified.
- I feel that `imgen` takes too many arguments: `temporary` is superfluous as it has to always be `True`; is it really necessary to pass `dirpath`, `prefix` and `suffix`— having to specify so many arguments doesn't defeat the purpose of this utility function (that is, to quickly save the image to an arbitrary file)?

**Comment by sugatoray** [comment]
@danoneata I kept `temporary` to allow for both permanently/temporarily saving an image. But in hindsight I feel it will be best to remove `temporary` and only handle temporary visualization here. 

I was using `quick_probe` to visualize the output after making any changes to the library -- just to make sure it is not broken, without having to provide any sample diagram to visualize.

**Comment by sugatoray** [comment]
> - I'm still not sold on the utility of the `create_sample_diagram` or the `quick_probe` functions; if we want to produce some default diagrams for testing purposes aren't the ones in the `examples` folder enough?

Consider this: Say, you are a user. And you installed the library from PyPI. You don't have the examples folder there. If you would like to create a demo diagram and play with it, with only a single function call here you get colored diagrams. It's an utility. It Is not meant to replace the need for importing `colour` and creating other diagrams that the user does later on. But alike many utilities in sklearn, where you could quickly download a dataset, this could quickly give you a sample diagram (or a set of diagrams) to explore various functionalities of the library.

> - I'm not sure about the logging part in `imgen` (it's a bit too verbose for my taste) and it's unclear whether adding a new dependency (`loguru`) just for this is justified.

`loguru` is imported as an optional dependency. So, if you prefer, it can be removed from the requirements and added as an optional dependency. 


> - I feel that `imgen` takes too many arguments: `temporary` is superfluous as it has to always be `True`; is it really necessary to pass `dirpath`, `prefix` and `suffix`— having to specify so many arguments doesn't defeat the purpose of this utility function (that is, to quickly save the image to an arbitrary file)?

The parameters `dirpath`, `prefix` and `suffix` are optional. So, in most cases the user will not need to specify these. With a known prefix/suffix you can spot a new temporary file created on your disk (otherwise it will have a random name).

**Comment by sugatoray** [comment]
@danoneata @srush 

I wish I had the scope managed down to only adding the display utilities here. 

Adding some docs led to adding more. 

Without those docstrings it was not be possible to create a PR for documentation website. And hence, until this PR is merged, I needed to add the contents of this PR into the other one. Not ideal. 

I know this could have been managed better. But thank you for your review, patience and accepting the PRs as is, although for the documentation PR, as @srush pointed out some pruning of contents -- Makefile, internal API docstrings, etc. would have been ideal.

**Pull request merged**